GPUDenseFiniteDifferenceImageFilter does not compile as shared library

A compile error occurs with both GPU and SHARED enabled. Example error message:

`itkGPUGradientAnisotropicDiffusionImageFilterTest.obj : error LNK2019: unresolved external symbol "public: static char const * __cdecl itk::GPUDenseFiniteDifferenceImageFilterKernel::GetOpenCLSource(void)" (?GetOpenCLSource@GPUDenseFiniteDifferenceImageFilterKernel@itk@@SAPEBDXZ) referenced in function "public: static char const * __cdecl itk::GPUDenseFiniteDifferenceImageFilter<class itk::GPUImage<float,2>,class itk::GPUImage<float,2>,class itk::GradientAnisotropicDiffusionImageFilter<class itk::GPUImage<float,2>,class itk::GPUImage<float,2> > >::GetOpenCLSource(void)" (?GetOpenCLSource@?$GPUDenseFiniteDifferenceImageFilter@V?$GPUImage@M$01@itk@@V12@V?$GradientAnisotropicDiffusionImageFilter@V?$GPUImage@M$01@itk@@V12@@2@@itk@@SAPEBDXZ)`

This problem was introduced by 4fe3b5de74c6114e907747758746b3c4ea4406ff.

### Steps to Reproduce

Turn on `BUILD_SHARED_LIBS` and `ITK_USE_GPU` during configure step.

### Reproducibility

100%

### Environment

Windows 10 x64, VS2019 & VS2017 (probably VS2015 too), multiple recent CMake versions.

--- a/CMake/itkOpenCL.cmake
+++ b/CMake/itkOpenCL.cmake
@@ -48,7 +48,7 @@ if(ITK_USE_GPU)
                       )
     set_source_files_properties(${CMAKE_CURRENT_BINARY_DIR}/${OUTPUT_FILE}
                                 PROPERTIES GENERATED ON)
-    set(${SRC_VAR} ${${SRC_VAR}} ${OUTPUT_FILE})
+    set(${SRC_VAR} ${${SRC_VAR}} ${CMAKE_CURRENT_BINARY_DIR}/${OUTPUT_FILE})
   endmacro()
 
   macro(write_gpu_kernels GPUKernels GPU_SRC)

--- a/Modules/Core/Common/include/itkMacro.h
+++ b/Modules/Core/Common/include/itkMacro.h
@@ -1146,22 +1146,16 @@ compilers.
  * provides the GPU kernel source code as a const char*
  */
 #define itkGPUKernelClassMacro(kernel)                                                                                 \
-  /**\class kernel \                                                                                                   \
-   * Workaround KWstyle bug \                                                                                          \
-   * \ingroup ITKCommon \                                                                                              \
-   */                                                                                                                  \
-  class kernel                                                                                                         \
+  kernel                                                                                                               \
   {                                                                                                                    \
   public:                                                                                                              \
-    static const char *                                                                                                \
-    GetOpenCLSource();                                                                                                 \
+    static const char * GetOpenCLSource();                                                                             \
                                                                                                                        \
   private:                                                                                                             \
     kernel();                                                                                                          \
     virtual ~kernel();                                                                                                 \
     kernel(const kernel &);                                                                                            \
-    void                                                                                                               \
-    operator=(const kernel &);                                                                                         \
+    void operator=(const kernel &);                                                                                    \
   };
 
 #define itkGetOpenCLSourceFromKernelMacro(kernel)                                                                      \

--- a/Modules/Core/GPUCommon/include/itkGPUImageOps.h
+++ b/Modules/Core/GPUCommon/include/itkGPUImageOps.h
@@ -23,7 +23,7 @@
 namespace itk
 {
 /** Create a helper GPU Kernel class for GPUImageOps */
-itkGPUKernelClassMacro(GPUImageOpsKernel);
+class itkGPUKernelClassMacro(GPUImageOpsKernel);
 
 /** \class GPUImageOps
  *

--- a/Modules/Core/GPUCommon/include/itkGPUReduction.h
+++ b/Modules/Core/GPUCommon/include/itkGPUReduction.h
@@ -26,7 +26,7 @@
 namespace itk
 {
 /** Create a helper GPU Kernel class for GPUReduction */
-itkGPUKernelClassMacro(GPUReductionKernel);
+class itkGPUKernelClassMacro(GPUReductionKernel);
 
 /**
  * \class GPUReduction

--- a/Modules/Core/GPUFiniteDifference/include/itkGPUDenseFiniteDifferenceImageFilter.h
+++ b/Modules/Core/GPUFiniteDifference/include/itkGPUDenseFiniteDifferenceImageFilter.h
@@ -24,7 +24,7 @@
 namespace itk
 {
 /** Create a helper GPU Kernel class for GPUDenseFiniteDifferenceImageFilter */
-itkGPUKernelClassMacro(GPUDenseFiniteDifferenceImageFilterKernel);
+class ITKGPUFiniteDifference_EXPORT itkGPUKernelClassMacro(GPUDenseFiniteDifferenceImageFilterKernel);
 
 /**
  * \class GPUDenseFiniteDifferenceImageFilter

--- a/Modules/Filtering/GPUAnisotropicSmoothing/include/itkGPUGradientNDAnisotropicDiffusionFunction.h
+++ b/Modules/Filtering/GPUAnisotropicSmoothing/include/itkGPUGradientNDAnisotropicDiffusionFunction.h
@@ -54,7 +54,7 @@ namespace itk
  */
 
 /** Create a helper GPU Kernel class for GPUGradientNDAnisotropicDiffusionFunction */
-itkGPUKernelClassMacro(GPUGradientNDAnisotropicDiffusionFunctionKernel);
+class itkGPUKernelClassMacro(GPUGradientNDAnisotropicDiffusionFunctionKernel);
 
 template <typename TImage>
 class ITK_TEMPLATE_EXPORT GPUGradientNDAnisotropicDiffusionFunction

--- a/Modules/Filtering/GPUAnisotropicSmoothing/include/itkGPUScalarAnisotropicDiffusionFunction.h
+++ b/Modules/Filtering/GPUAnisotropicSmoothing/include/itkGPUScalarAnisotropicDiffusionFunction.h
@@ -32,7 +32,7 @@ namespace itk
  * */
 
 /** Create a helper GPU Kernel class for GPUScalarAnisotropicDiffusionFunction */
-itkGPUKernelClassMacro(GPUScalarAnisotropicDiffusionFunctionKernel);
+class itkGPUKernelClassMacro(GPUScalarAnisotropicDiffusionFunctionKernel);
 
 template <typename TImage>
 class ITK_TEMPLATE_EXPORT GPUScalarAnisotropicDiffusionFunction : public GPUAnisotropicDiffusionFunction<TImage>

--- a/Modules/Filtering/GPUImageFilterBase/include/itkGPUCastImageFilter.h
+++ b/Modules/Filtering/GPUImageFilterBase/include/itkGPUCastImageFilter.h
@@ -29,7 +29,7 @@ namespace itk
 {
 
 /** Create a helper GPU Kernel class for GPUCastImageFilter */
-itkGPUKernelClassMacro(GPUCastImageFilterKernel);
+class itkGPUKernelClassMacro(GPUCastImageFilterKernel);
 
 /** \class GPUCastImageFilter
  * \brief GPU version of CastImageFilter.

--- a/Modules/Filtering/GPUImageFilterBase/include/itkGPUNeighborhoodOperatorImageFilter.h
+++ b/Modules/Filtering/GPUImageFilterBase/include/itkGPUNeighborhoodOperatorImageFilter.h
@@ -41,7 +41,7 @@ namespace itk
  */
 
 /** Create a helper GPU Kernel class for GPUNeighborhoodOperatorImageFilter */
-itkGPUKernelClassMacro(GPUNeighborhoodOperatorImageFilterKernel);
+class itkGPUKernelClassMacro(GPUNeighborhoodOperatorImageFilterKernel);
 
 template <typename TInputImage,
           typename TOutputImage,

--- a/Modules/Filtering/GPUSmoothing/include/itkGPUMeanImageFilter.h
+++ b/Modules/Filtering/GPUSmoothing/include/itkGPUMeanImageFilter.h
@@ -37,7 +37,7 @@ namespace itk
  */
 
 /** Create a helper GPU Kernel class for GPUMeanImageFilter */
-itkGPUKernelClassMacro(GPUMeanImageFilterKernel);
+class itkGPUKernelClassMacro(GPUMeanImageFilterKernel);
 
 template <typename TInputImage, typename TOutputImage>
 class ITK_TEMPLATE_EXPORT GPUMeanImageFilter

--- a/Modules/Filtering/GPUThresholding/include/itkGPUBinaryThresholdImageFilter.h
+++ b/Modules/Filtering/GPUThresholding/include/itkGPUBinaryThresholdImageFilter.h
@@ -85,7 +85,7 @@ private:
 } // end of namespace Functor
 
 /** Create a helper GPU Kernel class for GPUBinaryThresholdImageFilter */
-itkGPUKernelClassMacro(GPUBinaryThresholdImageFilterKernel);
+class itkGPUKernelClassMacro(GPUBinaryThresholdImageFilterKernel);
 
 /**
  * \class GPUBinaryThresholdImageFilter

--- a/Modules/Registration/GPUPDEDeformable/include/itkGPUDemonsRegistrationFunction.h
+++ b/Modules/Registration/GPUPDEDeformable/include/itkGPUDemonsRegistrationFunction.h
@@ -51,7 +51,7 @@ namespace itk
  * \ingroup ITKGPUPDEDeformableRegistration
  */
 /** Create a helper GPU Kernel class for GPUDemonsRegistrationFunction */
-itkGPUKernelClassMacro(GPUDemonsRegistrationFunctionKernel);
+class itkGPUKernelClassMacro(GPUDemonsRegistrationFunctionKernel);
 
 template <typename TFixedImage, typename TMovingImage, typename TDisplacementField>
 class ITK_TEMPLATE_EXPORT GPUDemonsRegistrationFunction

--- a/Modules/Registration/GPUPDEDeformable/include/itkGPUPDEDeformableRegistrationFilter.h
+++ b/Modules/Registration/GPUPDEDeformable/include/itkGPUPDEDeformableRegistrationFilter.h
@@ -73,7 +73,7 @@ namespace itk
  */
 
 /** Create a helper GPU Kernel class for GPUPDEDeformableRegistrationFilter */
-itkGPUKernelClassMacro(GPUPDEDeformableRegistrationFilterKernel);
+class itkGPUKernelClassMacro(GPUPDEDeformableRegistrationFilterKernel);
 
 template <typename TFixedImage,
           typename TMovingImage,

